File: fincept-terminal-desktop/src/services/news/newsMonitorService.ts
// News Monitor Service
// SQLite-backed keyword alert monitors for the News Tab.
// All persistence is via Tauri invoke (not localStorage).

import { invoke } from '@tauri-apps/api/core';
import { NewsArticle } from './newsService';

// ─── Types ────────────────────────────────────────────────────────────────────

export interface NewsMonitor {
  id: string;
  label: string;
  keywords: string[];
  color: string;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

// ─── Color palette (10 colors, cycles when adding monitors) ──────────────────

export const MONITOR_COLORS: string[] = [
  '#00E5FF', // cyan
  '#00FF88', // green
  '#FF6B35', // orange
  '#FFD700', // gold
  '#9D4EDD', // purple
  '#FF4D6D', // red-pink
  '#4CC9F0', // sky blue
  '#F72585', // hot pink
  '#7FFF00', // chartreuse
  '#FF9500', // amber
];

// ─── Helpers ──────────────────────────────────────────────────────────────────

function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function buildMatchRegex(keywords: string[]): RegExp | null {
  if (keywords.length === 0) return null;
  const pattern = keywords
    .map(k => `\\b${escapeRegex(k.trim())}\\b`)
    .join('|');
  try {
    return new RegExp(pattern, 'gi');
  } catch {
    return null;
  }
}

function articleMatchesMonitor(article: NewsArticle, monitor: NewsMonitor): boolean {
  if (!monitor.enabled || monitor.keywords.length === 0) return false;
  const regex = buildMatchRegex(monitor.keywords);
  if (!regex) return false;
  const text = `${article.headline} ${article.summary}`;
  regex.lastIndex = 0;
  return regex.test(text);
}

// ─── CRUD — all backed by SQLite via Tauri ───────────────────────────────────

export async function getMonitors(): Promise<NewsMonitor[]> {
  try {
    return await invoke<NewsMonitor[]>('get_news_monitors');
  } catch (e) {
    console.error('[NewsMonitorService] get_news_monitors failed:', e);
    return [];
  }
}

export async function addMonitor(
  label: string,
  keywords: string[],
  color?: string
): Promise<NewsMonitor | null> {
  try {
    // Auto-assign next color if not provided (caller can pass a custom one)
    const assignedColor = color ?? MONITOR_COLORS[0];
    return await invoke<NewsMonitor>('add_news_monitor', {
      label: label.trim(),
      keywords: keywords.map(k => k.trim()).filter(Boolean),
      color: assignedColor,
    });
  } catch (e) {
    console.error('[NewsMonitorService] add_news_monitor failed:', e);
    return null;
  }
}

export async function updateMonitor(
  id: string,
  label: string,
  keywords: string[],
  color: string,
  enabled: boolean
): Promise<boolean> {
  try {
    await invoke('update_news_monitor', {
      id,
      label: label.trim(),
      keywords: keywords.map(k => k.trim()).filter(Boolean),
      color,
      enabled,
    });
    return true;
  } catch (e) {
    console.error('[NewsMonitorService] update_news_monitor failed:', e);
    return false;
  }
}

export async function deleteMonitor(id: string): Promise<boolean> {
  try {
    await invoke('delete_news_monitor', { id });
    return true;
  } catch (e) {
    console.error('[NewsMonitorService] delete_news_monitor failed:', e);
    return false;
  }
}

export async function toggleMonitor(id: string, enabled: boolean): Promise<boolean> {
  try {
    await invoke('toggle_news_monitor', { id, enabled });
    return true;
  } catch (e) {
    console.error('[NewsMonitorService] toggle_news_monitor failed:', e);
    return false;
  }
}

// ─── Scanning ────────────────────────────────────────────────────────────────

/**
 * Scan articles against all monitors.
 * Returns a Map of monitorId → matched articles (de-duplicated by article id).
 */
export function scanMonitors(
  monitors: NewsMonitor[],
  articles: NewsArticle[]
): Map<string, NewsArticle[]> {
  const results = new Map<string, NewsArticle[]>();

  for (const monitor of monitors) {
    if (!monitor.enabled) continue;
    const matches: NewsArticle[] = [];
    const seen = new Set<string>();
    for (const article of articles) {
      if (!seen.has(article.id) && articleMatchesMonitor(article, monitor)) {
        matches.push(article);
        seen.add(article.id);
      }
    }
    results.set(monitor.id, matches);
  }

  return results;
}

/**
 * Detect new monitor matches that weren't in the previous fetch.
 * Used to fire notifications only for newly-arrived articles.
 *
 * @param monitors    All active monitors
 * @param articles    Current article list
 * @param prevIds     Set of article IDs from the previous refresh cycle
 * @returns           Array of { monitor, article } pairs to notify about
 */
export function checkForNewBreakingMatches(
  monitors: NewsMonitor[],
  articles: NewsArticle[],
  prevIds: Set<string>
): { monitor: NewsMonitor; article: NewsArticle }[] {
  const results: { monitor: NewsMonitor; article: NewsArticle }[] = [];

  const newArticles = articles.filter(
    a => !prevIds.has(a.id) && (a.priority === 'FLASH' || a.priority === 'URGENT')
  );

  for (const monitor of monitors) {
    if (!monitor.enabled) continue;
    for (const article of newArticles) {
      if (articleMatchesMonitor(article, monitor)) {
        results.push({ monitor, article });
      }
    }
  }

  return results;
}

/**
 * Determine which color to assign the next monitor based on existing monitors.
 */
export function nextMonitorColor(existingMonitors: NewsMonitor[]): string {
  const usedColors = new Set(existingMonitors.map(m => m.color));
  return (
    MONITOR_COLORS.find(c => !usedColors.has(c)) ??
    MONITOR_COLORS[existingMonitors.length % MONITOR_COLORS.length]
  );
}

File: fincept-qt/src/services/news/NewsMonitorService.cpp
#include "services/news/NewsMonitorService.h"

#include "core/logging/Logger.h"
#include "storage/repositories/NewsMonitorRepository.h"

#include <QRegularExpression>
#include <QUuid>

namespace fincept::services {

const QStringList& monitor_colors() {
    static const QStringList colors = {
        "#00E5FF", "#00FF88", "#FF6B35", "#FFD700", "#9D4EDD", "#FF4D6D", "#4CC9F0", "#F72585", "#7FFF00", "#FF9500",
    };
    return colors;
}

NewsMonitorService& NewsMonitorService::instance() {
    static NewsMonitorService s;
    return s;
}

NewsMonitorService::NewsMonitorService() {
    // Table now created by MigrationRunner v001 — no ensure_table() needed
    load_from_db();
}

void NewsMonitorService::load_from_db() {
    monitors_.clear();
    auto result = fincept::NewsMonitorRepository::instance().list_all();
    if (result.is_err())
        return;
    for (const auto& row : result.value()) {
        NewsMonitor m;
        m.id = row.id;
        m.label = row.label;
        m.keywords = row.keywords;
        m.color = row.color;
        m.enabled = row.enabled;
        monitors_.append(m);
    }
}

void NewsMonitorService::save_to_db(const NewsMonitor& m) {
    fincept::NewsMonitorRow row{m.id, m.label, m.keywords, m.color, m.enabled};
    fincept::NewsMonitorRepository::instance().upsert(row);
}

void NewsMonitorService::remove_from_db(const QString& id) {
    fincept::NewsMonitorRepository::instance().remove(id);
}

QVector<NewsMonitor> NewsMonitorService::get_monitors() const {
    return monitors_;
}

void NewsMonitorService::add_monitor(const QString& label, const QStringList& keywords, const QString& color) {
    NewsMonitor m;
    m.id = QUuid::createUuid().toString(QUuid::WithoutBraces).left(8);
    m.label = label;
    m.keywords = keywords;
    m.color = color.isEmpty() ? monitor_colors()[monitors_.size() % monitor_colors().size()] : color;
    m.enabled = true;
    monitors_.append(m);
    save_to_db(m);
}

void NewsMonitorService::update_monitor(const NewsMonitor& monitor) {
    for (auto& m : monitors_) {
        if (m.id == monitor.id) {
            m = monitor;
            save_to_db(m);
            return;
        }
    }
}

void NewsMonitorService::delete_monitor(const QString& id) {
    monitors_.erase(
        std::remove_if(monitors_.begin(), monitors_.end(), [&](const NewsMonitor& m) { return m.id == id; }),
        monitors_.end());
    remove_from_db(id);
}

void NewsMonitorService::toggle_monitor(const QString& id) {
    for (auto& m : monitors_) {
        if (m.id == id) {
            m.enabled = !m.enabled;
            save_to_db(m);
            return;
        }
    }
}

QMap<QString, QVector<NewsArticle>> NewsMonitorService::scan_monitors(const QVector<NewsMonitor>& monitors,
                                                                      const QVector<NewsArticle>& articles) const {
    QMap<QString, QVector<NewsArticle>> result;

    for (const auto& mon : monitors) {
        if (!mon.enabled || mon.keywords.isEmpty())
            continue;

        // Build regex from keywords
        QStringList escaped;
        for (const auto& kw : mon.keywords) {
            escaped << ("\\b" + QRegularExpression::escape(kw.trimmed()) + "\\b");
        }
        QRegularExpression re(escaped.join("|"), QRegularExpression::CaseInsensitiveOption);
        if (!re.isValid())
            continue;

        QVector<NewsArticle> matches;
        for (const auto& a : articles) {
            QString text = a.headline + " " + a.summary;
            if (re.match(text).hasMatch())
                matches.append(a);
        }
        if (!matches.isEmpty())
            result[mon.id] = matches;
    }

    return result;
}

} // namespace fincept::services
